HIP/hip-0000-introduction-of-blocks.md

```diff
@@ -0,0 +1,199 @@
+
+## Specification
+
+**Definitions**
+
+- `block` → `Record file` containing all `Record Stream Objects` grouped in a 2-second time frame.
+- `block number` → consecutive number of the `Record file` that is being incremented by `1` for every new `Record file`. This value can be initially bootstrapped through `Mirror Nodes` and after that maintained by services nodes.
+- `block hash` → The `32 byte keccak256 running hash` of the `Record file`. That is the running hash of the last `Record Stream Object` from the previous `Record file`.
+- `block timestamp` → Instant of consensus timestamp of the first `transaction`/`Record Stream Object` in the `Record file`.
+
+### Platform
+
+Adapt `TimestampStreamFileWriter` to include `blockNumber` in the `Record File`. Introduce a new field `firstConsensusTimeInCurrentFile` to be used as a marker when to start a new Record file. Use the field `lastConsensusTimestamp` to keep track of the last-seen consensus timestamp that was processed. In this way, we can ensure that we have at least `1000ns`  difference between the last processed transaction before a new file is created. In this way if we have a `parent transaction` with at least one `child precompile transaction` they all will be included in the same `block`/`Record file`. Otherwise, we might have a corner case where a `parent transaction` is included in one block, and its `child precompile transaction` falls into the next block since it will increase the `consensus timestamp` with `1ns`. Therefore, the algorithm for checking whether to start a new file would be the following:
+
+A new `Record Stream Object` enters `addObject(T object)` in `TimestampStreamFileWriter`. It has a consensus timestamp `T`. We create a new file only if both (1) and (2) conditions are met:
+
+1. `T - lastConsensusTime > 1000ns`
+2. `T - firstConsensusTimeInCurrentFile > 2s`
+
+or
+
+if `lastConsensusTime` or `firstConsensusTimeInCurrentFile` is `null`
+
+### Services
+
+Services is to update the processing logic of transactions so that it supports logic for determining new 2-second periods, increment `block` number and keep `block` relevant data.
+The proposed solution specifies a `long` field to be used for the block number counter, incremented every 2 seconds.
+
+**Pseudo-code of the record streaming algorithm**
+
+```markdown
+Properties in State:
+- blockHashes `map(number -> keccak256(RunningHash))` - stores the hashes of the last 256 blocks
+- blockNumber `long` - stores the current block number
+- blockTimestamp - `Instant` - stores the timestamp of the block
+
+handleTransaction() {
+	...
+	bool `newBlock` = shouldCreateNewBlock() {
+			if (`currentTS` - `lastConsensusTime` > `1000ns` && `currentTS` - `blockTimestamp` > `2s`) return `true`; else `false`
```

Comment:
`>=` - greater than or equal? Either this would cause 3 second blocks or a slow drift away from two seconds.  Perhaps express it as a modulus of the preferred block time? Or we consider the floors of the current and block timestamp to be >= 2?

Reply:
We are considering that the minimum period of time for a block to be created is `hedera.recordStream.logPeriod` seconds, which currently is set to `2 seconds`. If there is no transaction coming in, record files are not getting exported, thus blocks are not created.

I am not sure that I understand the first part of your comment.